src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -1991,29 +1996,438 @@ static void Main()
+                // (2,10): error CS8773: Feature 'inferred delegate type' is not available in C# 9.0. Please use language version 10.0 or greater.
+                // var d2 = () => { };
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion9, "() => { }").WithArguments("inferred delegate type", "10.0").WithLocation(2, 10),
+                // (3,10): error CS8773: Feature 'inferred delegate type' is not available in C# 9.0. Please use language version 10.0 or greater.
+                // var d3 = delegate () { };
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion9, "delegate () { }").WithArguments("inferred delegate type", "10.0").WithLocation(3, 10));
+
+            comp = CreateCompilation(source, parseOptions: TestOptions.Regular10.WithKind(SourceCodeKind.Script));
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ImplicitlyTypedVariables_03()
+        {
+            var source =
+@"class Program
+{
+    static void Main()
+    {
+        ref var d1 = Main;
+        ref var d2 = () => { };
+        ref var d3 = delegate () { };
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (5,17): error CS8172: Cannot initialize a by-reference variable with a value
+                //         ref var d1 = Main;
+                Diagnostic(ErrorCode.ERR_InitializeByReferenceVariableWithValue, "d1 = Main").WithLocation(5, 17),
+                // (5,22): error CS1657: Cannot use 'Main' as a ref or out value because it is a 'method group'
+                //         ref var d1 = Main;
+                Diagnostic(ErrorCode.ERR_RefReadonlyLocalCause, "Main").WithArguments("Main", "method group").WithLocation(5, 22),
+                // (6,17): error CS8172: Cannot initialize a by-reference variable with a value
+                //         ref var d2 = () => { };
+                Diagnostic(ErrorCode.ERR_InitializeByReferenceVariableWithValue, "d2 = () => { }").WithLocation(6, 17),
+                // (6,22): error CS1510: A ref or out value must be an assignable variable
+                //         ref var d2 = () => { };
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "() => { }").WithLocation(6, 22),
+                // (7,17): error CS8172: Cannot initialize a by-reference variable with a value
+                //         ref var d3 = delegate () { };
+                Diagnostic(ErrorCode.ERR_InitializeByReferenceVariableWithValue, "d3 = delegate () { }").WithLocation(7, 17),
+                // (7,22): error CS1510: A ref or out value must be an assignable variable
+                //         ref var d3 = delegate () { };
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "delegate () { }").WithLocation(7, 22));
+        }
+
+        [Fact]
+        public void ImplicitlyTypedVariables_04()
```

Comment:
>ImplicitlyTypedVariables_04 [](http://example.com/codeflow?start=20&length=27)

Would it be worth making a version of this test where `Action` inherits from `IDisposable`?